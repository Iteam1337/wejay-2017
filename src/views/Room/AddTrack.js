import React, { Component } from 'react'
import { roomQuery } from './Room'
import { gql, graphql } from 'react-apollo'
import md5 from 'md5'

type AddTrackProps = {
  mutate: Function,
  roomName: string
}

export class AddTrack extends Component {
  props: AddTrackProps

  handleKeyUp = evt => {
    const { mutate, roomName } = this.props

    if (evt.keyCode === 13) {
      const userId = md5(localStorage.getItem('user'))

      mutate({
        variables: {
          input: {
            roomName,
            userId,
            spotifyId: evt.target.value
          }
        },
        optimisticResponse: {
          queueTrack: {
            album: [],
            artists: [],
            duration: 0,
            name: '',
            spotifyUri: evt.target.value,
            user: '',
            __typename: 'Track'
          }
        },
        update: (store, { data: { queueTrack } }) => {
          const data = store.readQuery({
            query: roomQuery,
            variables: {
              name: roomName
            }
          })

          const exists = data.room.queue.find(
            track => track.spotifyUri === queueTrack.spotifyUri
          )

          const notCurrent =
            data.room.currentTrack &&
            data.room.currentTrack.spotifyUri !== queueTrack.spotifyUri

          if (!exists && notCurrent) {
            data.room.queue.push(queueTrack)
          }

          store.writeQuery({
            query: roomQuery,
            variables: {
              name: roomName
            },
            data
          })
        }
      })

      evt.target.value = ''
    }
  }

  render () {
    return (
      <div>
        <input
          onKeyUp={this.handleKeyUp}
          placeholder="Spotify URI"
          type="text"
        />
      </div>
    )
  }
}

const addTrackMutation = gql`
  mutation queueTrack($input: QueueInput!) {
    queueTrack(input: $input) {
      album {
        images {
          url
        }
        name
      }
      artists {
        name
      }
      duration
      name
      spotifyUri
      user
    }
  }
`

export default graphql(addTrackMutation)(AddTrack)
