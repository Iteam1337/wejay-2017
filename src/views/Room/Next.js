import React from 'react'
import { gql, graphql } from 'react-apollo'

export const Next = ({ mutate, roomName }) => {
  const handleNext = () => {
    mutate({
      variables: {
        roomName
      }
    })
  }

  return <button onClick={handleNext}>Next</button>
}

const nextMutation = gql`
  mutation nextTrack($roomName: String!) {
    nextTrack(roomName: $roomName) {
      name
    }
  }
`

export default graphql(nextMutation)(Next)
