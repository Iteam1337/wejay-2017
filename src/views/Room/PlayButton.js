// @flow

import React, { Component } from 'react'
import { compose, graphql } from 'react-apollo'
import gql from 'graphql-tag'
import Icon from 'react-icons-kit'
import { play } from 'react-icons-kit/ikons/play'
import { pause } from 'react-icons-kit/ikons/pause'
import styled from 'styled-components'

type PlayButtonProps = {
  isPlaying: boolean,
  pause: Function,
  play: Function,
  roomName: string,
}

const PlayButtonWrap = styled.button`
  -moz-appearance: none;
  -webkit-appearance: none;
  align-items: center;
  background-color: ${({ theme }) => theme.colors.primary};
  border: 0;
  border-radius: 100%;
  color: #fff;
  display: flex;
  height: 80px;
  justify-content: center;
  width: 80px;

  &:focus {
    outline: none;
  }
`

class PlayButton extends Component<PlayButtonProps> {
  handleClick = () => {
    const { isPlaying, pause, play, roomName } = this.props
    const variables = {
      roomName,
    }

    if (isPlaying) {
      pause({ variables })
    } else {
      play({ variables })
    }
  }

  render () {
    return (
      <PlayButtonWrap onClick={this.handleClick}>
        <Icon icon={this.props.isPlaying ? pause : play} size={60} />
      </PlayButtonWrap>
    )
  }
}

const pauseSubscription = gql`
  mutation pause($roomName: String!) {
    pause(roomName: $roomName)
  }
`

const playSubscription = gql`
  mutation play($roomName: String!) {
    play(roomName: $roomName)
  }
`

export default compose(
  graphql(pauseSubscription, { name: 'pause' }),
  graphql(playSubscription, { name: 'play' })
)(PlayButton)
