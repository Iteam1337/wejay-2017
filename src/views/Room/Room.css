.Room {
  display: grid;
  grid-template-columns: 300px 1fr;
  grid-column-gap: 30px;
  margin: 120px auto 0;
  width: 80vw;
  position: relative;
  z-index: 10; }

.Room__overlay {
  background: rgba(36, 41, 46, 0.95);
  bottom: 0;
  left: 0;
  position: absolute;
  right: 0;
  top: 0; }

.Room__content {
  display: grid;
  grid-row-gap: 30px; }
