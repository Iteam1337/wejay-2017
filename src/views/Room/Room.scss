.Room {
  align-items: flex-start;
  display: grid;
  grid-template-columns: 300px 1fr 300px;
  grid-column-gap: 60px;
  margin: 120px auto 0;
  width: 80vw;
  position: relative;
  z-index: 10;
}

.Room__overlay {
  background: rgba(36, 41, 46, 0.95);
  bottom: 0;
  left: 0;
  position: absolute;
  right: 0;
  top: 0;
}

.Room__title {
  margin-bottom: 40px;
}

.Room__now-playing {
  align-items: center;
  display: flex;
  margin-top: 40px;
}

.Room__now-playing-track {
  flex: 1;
}
