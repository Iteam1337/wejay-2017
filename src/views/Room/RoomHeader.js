// @flow

import React from 'react'
import styled from 'styled-components'
import humanizeDuration from 'humanize-duration'
import type { RoomType } from './Room'
import pluralize from 'pluralize'

type RoomHeaderProps = {
  name: string,
  queue: $PropertyType<RoomType, 'queue'>,
  users: number,
}

const RoomHeaderWrap = styled.div``
const PlaylistTitle = styled.div`
  font-size: 12px;
  text-transform: uppercase;
`

const RoomName = styled.div`
  color: rgb(237, 167, 76);
  font-family: 'Roboto Slab';
  font-size: 32px;
  font-weight: 400;
  line-height: 1.5;
`

const RoomMeta = styled.div`
  color: rgba(54, 61, 67, 1);
  font-family: 'Roboto Slab';
  font-size: 14px;
  font-weight: 400;
`

const RoomHeader = ({ name, queue, users }: RoomHeaderProps) => {
  const duration = queue.reduce((acc, track) => acc + track.duration, 0)

  return (
    <RoomHeaderWrap>
      <PlaylistTitle>Playlist title</PlaylistTitle>
      <RoomName>{name}</RoomName>
      <RoomMeta>
        {queue.length} {pluralize('track', queue.length)} •{' '}
        {humanizeDuration(duration, { delimiter: ' - ', round: true })} •{' '}
        {users} {pluralize('contributor', users)}
      </RoomMeta>
    </RoomHeaderWrap>
  )
}

export default RoomHeader
