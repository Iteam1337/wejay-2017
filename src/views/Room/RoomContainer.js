// @flow

import * as Wejay from 'typings/wejay.flow'
import * as WejayApi from './__generated__/RoomQuery'
import React, { createContext, Component } from 'react'
import Room from './Room'
import gql from 'graphql-tag'
import { compose, graphql } from 'react-apollo'
import * as storage from '../../utils/storage'

type RoomContainerProps = {
  addTrack: (options: Object) => Promise<void>,
  data: Wejay.ApolloBase<WejayApi.RoomQuery>,
  match: {
    params: {
      name: string,
    },
  },
}

export const RoomContext = createContext({
  queue: [],
})

class RoomContainer extends Component<RoomContainerProps> {
  componentDidMount () {
    const roomName = this.props.match.params.name

    this.props.data.subscribeToMore({
      document: RoomUpdated,
      variables: {
        roomName,
      },
      updateQuery: (prev, { subscriptionData }) => {
        if (!subscriptionData.data) {
          return prev
        }

        return {
          ...prev,
          room: {
            ...prev.room,
            ...subscriptionData.data.roomUpdated,
          },
        }
      },
    })
  }

  addToQueue = async spotifyId => {
    const roomName = this.props.match.params.name
    const savedId = localStorage.getItem('id')
    const { id } = await storage.getValue(savedId)

    await this.props.addTrack({
      variables: {
        input: {
          roomName,
          userId: id,
          spotifyId,
        },
      },
      optimisticResponse: {
        queueTrack: {
          album: {
            images: [],
            name: 'Fake album',
            __typename: 'Album',
          },
          artists: [
            {
              name: 'Loading',
              __typename: 'Artist',
            },
          ],
          duration: 0,
          name: 'Awesome track',
          spotifyUri: spotifyId,
          started: 0,
          user: {
            id,
            email: '',
            __typename: 'User',
          },
          __typename: 'Track',
        },
      },
      update: (store, { data: { queueTrack } }) => {
        const data = store.readQuery({
          query: RoomQuery,
          variables: {
            name: roomName,
          },
        })

        const exists = data.room.queue.find(
          track => track.spotifyUri === queueTrack.spotifyUri
        )

        const notCurrent =
          data.room.currentTrack &&
          data.room.currentTrack.spotifyUri !== queueTrack.spotifyUri

        if (!exists && notCurrent) {
          data.room.queue.push(queueTrack)
        }

        store.writeQuery({
          query: RoomQuery,
          variables: {
            name: roomName,
          },
          data,
        })
      },
    })
  }

  render () {
    const { data: { error, loading, room } } = this.props

    if (loading) {
      return (
        <div className="Loading">
          Loading
          <div className="Loader" />
        </div>
      )
    }

    if (error) {
      return <div>{error.message}</div>
    }

    return (
      <RoomContext.Provider
        value={{
          room,
          actions: { addToQueue: this.addToQueue },
        }}
      >
        <Room addToQueue={this.addToQueue} room={room} />
      </RoomContext.Provider>
    )
  }
}

export const TrackInfoFragment = gql`
  fragment TrackInfo on Track {
    album {
      images {
        height
        url
        width
      }
      name
    }
    artists {
      name
    }
    duration
    name
    spotifyUri
    started
    user {
      email
      id
    }
  }
`

const UserInfoFragment = gql`
  fragment UserInfo on User {
    email
    id
    lastPlay
  }
`

export const RoomQuery = gql`
  query RoomQuery($name: String!) {
    room(name: $name) {
      currentTrack {
        ...TrackInfo
      }
      isPlaying
      name
      users {
        ...UserInfo
      }
      queue {
        ...TrackInfo
      }
    }
  }

  ${TrackInfoFragment}
  ${UserInfoFragment}
`

export const RoomUpdated = gql`
  subscription RoomUpdated($roomName: String!) {
    roomUpdated(roomName: $roomName) {
      currentTrack {
        ...TrackInfo
      }
      isPlaying
      name
      users {
        ...UserInfo
      }
      queue {
        ...TrackInfo
      }
    }
  }

  ${TrackInfoFragment}
  ${UserInfoFragment}
`

const AddTrackMutation = gql`
  mutation QueueTrack($input: QueueInput!) {
    queueTrack(input: $input) {
      album {
        images {
          url
        }
        name
      }
      artists {
        name
      }
      duration
      name
      spotifyUri
      user {
        email
        id
      }
    }
  }
`

export default compose(
  graphql(RoomQuery, {
    options: props => ({
      variables: { name: props.match.params.name },
    }),
  }),
  graphql(AddTrackMutation, { name: 'addTrack' })
)(RoomContainer)
