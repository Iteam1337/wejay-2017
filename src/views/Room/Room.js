// @flow

import './Room.css'
import React, { Component } from 'react'
import Queue from 'components/Queue'
import Track from 'components/Track'
import AddTrack from './AddTrack'
import Next from './Next'
import { gql, graphql } from 'react-apollo'
import Gravatar from 'components/Gravatar'
import Droparea from 'components/Droparea'

export class Room extends Component {
  componentWillMount () {
    this.props.data.subscribeToMore({
      document: queueUpdated,
      variables: {
        roomName: this.props.match.params.name
      },
      updateQuery: (prev, { subscriptionData }) => {
        if (!subscriptionData.data) {
          return prev
        }

        return {
          ...prev,
          room: {
            ...prev.room,
            queue: subscriptionData.data.queueUpdated
          }
        }
      }
    })

    this.props.data.subscribeToMore({
      document: onNextTrack,
      variables: {
        roomName: this.props.match.params.name
      },
      updateQuery: (prev, { subscriptionData }) => {
        if (!subscriptionData.data) {
          return prev
        }

        return {
          ...prev,
          room: {
            ...prev.room,
            currentTrack: subscriptionData.data.onNextTrack
          }
        }
      }
    })
  }

  render () {
    const { data: { error, loading, room } } = this.props

    if (loading) {
      return <div>Loading</div>
    }

    if (error) {
      return <div>{error.message}</div>
    }

    return (
      <div>
        <Droparea roomName={room.name} />
        <div className="Room">
          <AddTrack roomName={room.name} />
          <Next roomName={room.name} />

          <Track track={room.currentTrack} />

          <ul>
            {room.users.map(user => (
              <li key={user.id}>
                <Gravatar email={user.email} />
              </li>
            ))}
          </ul>

          <Queue tracks={room.queue} />
        </div>
      </div>
    )
  }
}

Room.fragments = {
  track: gql`
    fragment TrackInfo on Track {
      album {
        images {
          url
        }
        name
      }
      artists {
        name
      }
      duration
      name
      spotifyUri
      user {
        email
        id
      }
    }
  `
}

export const roomQuery = gql`
  query RoomQuery($name: String!) {
    room(name: $name) {
      currentTrack {
        ...TrackInfo
      }
      name
      users {
        email
        id
      }
      queue {
        ...TrackInfo
      }
    }
  }
  ${Room.fragments.track}
`

const queueUpdated = gql`
  subscription queueUpdated($roomName: String!) {
    queueUpdated(roomName: $roomName) {
      ...TrackInfo
    }
  }
  ${Room.fragments.track}
  `

const onNextTrack = gql`
  subscription onNextTrack($roomName: String!) {
    onNextTrack(roomName: $roomName) {
      ...TrackInfo
    }
  }
  ${Room.fragments.track}
`

export default graphql(roomQuery, {
  options: props => ({
    variables: { name: props.match.params.name }
  })
})(Room)
