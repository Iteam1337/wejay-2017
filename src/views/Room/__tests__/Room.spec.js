import React from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'
import { Room } from '../Room'

describe('components/Room', () => {
  let component
  let data

  beforeEach(() => {
    data = {
      error: null,
      loading: false,
      room: {
        currentTrack: {
          artists: [{ name: 'artist' }, { name: 'artist two' }],
          name: 'track',
          user: {
            id: '1337'
          }
        },
        name: 'test',
        queue: [],
        users: []
      },
      subscribeToMore: jest.fn()
    }

    component = shallow(
      <Room
        data={data}
        match={{
          params: {
            name: 'test'
          }
        }}
      />
    )
  })

  it('renders a loading message', () => {
    component.setProps({
      data: {
        ...data,
        loading: true
      }
    })

    expect(toJson(component)).toMatchSnapshot()
  })

  it('renders error message', () => {
    component.setProps({
      data: {
        ...data,
        error: {
          message: 'Error'
        }
      }
    })

    expect(toJson(component)).toMatchSnapshot()
  })

  it('renders Room', () => {
    expect(toJson(component)).toMatchSnapshot()
  })
})
