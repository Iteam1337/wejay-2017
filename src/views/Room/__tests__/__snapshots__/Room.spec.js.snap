// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Room renders Room 1`] = `
<div>
  <Apollo(Droparea)
    roomName="test"
  />
  <div
    className="Room"
  >
    <div
      className="Room__cover"
    >
      <h2
        className="Room__title"
      >
        Now Playing
      </h2>
      <Component
        track={
          Object {
            "artists": Array [
              Object {
                "name": "artist",
              },
              Object {
                "name": "artist two",
              },
            ],
            "name": "track",
            "user": Object {
              "id": "1337",
            },
          }
        }
      />
      <div
        className="Room__now-playing"
      >
        <div
          className="Room__now-playing-track"
        >
          <div
            className="Track__artist"
          >
            artist, artist two
          </div>
          <div
            className="Track__name"
          >
            track
          </div>
        </div>
        <Gravatar
          email=""
          id="1337"
          size={30}
        />
      </div>
    </div>
    <div
      className="Room__content"
    >
      <h2
        className="Room__title"
      >
        Queue
      </h2>
      <Queue
        tracks={Array []}
      />
    </div>
    <div>
      <h2
        className="Room__title"
      >
        Users
      </h2>
      <Users
        users={Array []}
      />
    </div>
  </div>
</div>
`;

exports[`components/Room renders a loading message 1`] = `
<div>
  Loading
</div>
`;

exports[`components/Room renders error message 1`] = `
<div>
  Error
</div>
`;
